### PR TITLE
python3 updates Bio.SeqIO.AbiIO

### PR DESCRIPTION
**Pull request by mdehoon**
This pull request updates Bio.SeqIO.AbiIO, and allows for other types of byte streams (e.g. BytesIO) which do not have a `mode` attribute.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

File: Bio/SeqIO/AbiIO.py [hunk 1]
```diff
@@ -360,15 +359,16 @@ def AbiIterator(handle, alphabet=None, trim=False):
...
-    # raise exception if handle mode is not 'rb'
-    if hasattr(handle, "mode"):
-        if set("rb") != set(handle.mode.lower()):
-            raise ValueError("ABI files has to be opened in 'rb' mode.")
+    try:
+        handle = open(argument, "rb")
+    except TypeError:
+        handle = argument
+        if handle.read(0) != b"":
+            raise ValueError("ABI files has to be opened in 'rb' mode.") from None
```
Comment (peterjc) [inline Bio/SeqIO/AbiIO.py]: Pre-existing typo, has/have.

Other file formats use something like "ABI files must be opened in binary mode." but we're not being entirely consistent here.

**Comment by peterjc** [comment]
Using ``source`` is a better argument name than ``argument`` as in the initial commit, but as on #2594 this is technically an API change for minimal benefit. Why not leave it as ``handle``?

**Comment by peterjc** [comment]
Rebased for the black problem on the master branch.

**Comment by mdehoon** [comment]
> Using `source` is a better argument name than `argument` as in the initial commit, but as on #2594 this is technically an API change for minimal benefit. Why not leave it as `handle`?

`handle` suggests file handles only.

**Comment by peterjc** [comment]
Yes - it is a trade off. Given you want to make this API (and others like it) more flexible by accepting more than just a handle - is worth breaking full backward compatibility for a slight improvement in the documentation clarity?

In the case of the low level SeqIO internal iterator, I think the change has very low risk. Extremely low in this case, how often do we expect to see people writing this  as ``AbiIterator(handle=XXX)`` versus the more compact ``AbiItertor(XXX)``? Almost never.

My concern is more flagging this for more widely used functionality, particularly code Python beginners are likely to use, where that risk is higher.

**Comment by mdehoon** [comment]
> Given you want to make this API (and others like it) more flexible by accepting more than just a handle - is worth breaking full backward compatibility for a slight improvement in the documentation clarity?

Absolutely. Maintaining full backward compatibility makes things easier for current users, but only if they were actually using the keyword, and if so it's a trivial fix. Keeping documentation unclear, however, will disadvantage all future Biopython users for all eternity.

**Comment by mdehoon** [comment]
@peterjc  Thanks! Let's wait and see if @bow has any comments before we merge this

**Pull request merged**